### DISCUSSION TITLE
Need help with first metainfo.xml for Tuxedo IBS 15 Gen6

### DISCUSSION THREAD

**Discussion by Matombo** [fwupd/fwupd#5772] [Q&A]
I successfully build a minimal metainfo.xml and flashed the BIOS chip of a Tuxedo IBS 15 Gen6/NS50MU.

However in doing so, all the custom Tuxedo dmi strings and the Tuxedo boot logo where gone and I had to manually restore them.

I now looked further into how the efi shell update works -> It is actually 2 stage:
1. from the 16MB firmware image only the first 4MB actually get flashed
2. After that a efi binary is executed that resets the device
3. It automatically boots to a non interactive efi environment and persumably finishes the flash by copying the 4MB from the beginning of the flash chip to the actual location. You see some meta information (barebone name, BIOS version) and a progress bar on screen.
I assume this is done to prevent a bricked device in case of a power loss, something like "A-B-flashing-light".

So my immediate questions:
1. Can I tell fwupd using the flashrom plugin to only flash 4MB instead of the full 16MB? I tried to cut the 16MB image and pack the new smaller image in the cab file, but then fwupd complains about a size missmatch.
2. Can fwupd reboot into executing the efi binary, that then reboots into the "2nd-stage-flash" mode? Or do I need to find out what this binary actually does so that it can be reimplemented in fwupd directly?

**Comment by hughsie** [comment]
I think we'd need some more info on how the EFI binary work; my gut instinct is that maybe we could set a quirk and truncate the image but then you have two hacks, not one.

**Reply by superm1** [reply]
I don't think we want to set precedent for flashrom plugin to be calling into random EFI binaries.  We (effectively) do that with UEFI runtime services `UpdateCapsule` call specifically because it works with UEFI secure boot, and the firmware has a good indication that such a runtime service exists.

IMO it would be better to learn how the binary works and we either port that to fwupdx64.efi and call that or we do more things from Linux to mimic what it does.

**Comment by Matombo** [comment]
For reference, the BIOS with the efi flash scripts: https://cloud.tuxedo.de/index.php/s/aMaPeJwTq74b6zr (pw: fwupd)

Short explanation of the relevant parts:
- FLASHME.nsh -> the device comes in a refresh and a non-refresh variant with different bioses in 68C1 and 68C2, this automatically selects the right folder
- 68C1/NS50_70MU.16NRTR4 -> the actuall BIOS rom image, size 16MB == physical size of the flash chip
- 68C1/Fpt.efi the efi flashtool for the BIOS chip
- 68C1/NS50_70MU.16NRTR4_tux.efi the efi binary that is executed after the 4MB of the image got flashed, it is itself even biger then the BIOS image and does seems to contain it
- 68C1/FLASHME.nsh the script to execut to perform a bios update via the efi shell
The rest seems to be optional and not related to the actuall flash process

The Progress bar screen also says "InsydeH2OFFT (Flash Firmware Tool) Version [...]" so it seems to not be something clevo specific, but general for all Insyde BIOSes. Maybe there are already other devices using an Insyde BIOS in LVFS and i can copy their approach?

Speaking of which: Steamdeck has the same bios flash progress bar screen, as an example for a non clevo device. And Steamdeck does trigger it from linux, so there is a way, just finding out how...

When running the binary without previously flashing a new BIOS, on reboot I get the message:
```
InsydeH2O - Secure Flash
Error: Invalid Firmware Image!!!
Please press any key to restart system.....
```
It works just fine afterwards

Just for reference because auf the name "InsydeH2O - Secure Flash"

**Reply by superm1** [reply]
> Just for reference because auf the name "InsydeH2O - Secure Flash"

Why not just use UEFI capsule?  It's a much clearer line about where the free software and proprietary software is used.  Once `UpdateCapsule` is called it's all IBV code.

IMO I think you'll need to talk to Inysde about improving this experience from fwupd.  I don't think mixing and matching parts of flashrom and proprietary tools is a good idea.

> Speaking of which: Steamdeck has the same bios flash progress bar screen, as an example for a non clevo device. And Steamdeck does 

Steamdeck doesn't use fwupd / LVFS IIRC.  I would anticipate they have a proprietary flasher.

> I can trigger it from linux, so there is a way, just finding out how...

The progressbar showing up doesn't tell you much really.
I can trigger it on an Insyde based device too from fwupd just by invoking a UEFI capsule update from runtime services.

**Reply by Matombo** [reply]
> > Just for reference because auf the name "InsydeH2O - Secure Flash"
> 
> Why not just use UEFI capsule? It's a much clearer line about where the free software and proprietary software is used. Once `UpdateCapsule` is called it's all IBV code.

I'm still figuring out the relevant formats and stuff around fwupd, just startet to look into it last week. For the moment I was just working with the files Clevo usually provides us with.

I will try to find a source to further educate myself.
> 
> IMO I think you'll need to talk to Inysde about improving this experience from fwupd. I don't think mixing and matching parts of flashrom and proprietary tools is a good idea.
Sadly we (aka TUXEDO Computers) have no direct contact to Insyde but only to Clevo and other barebone/mainboard manufacturers, we might be able to make a request for upcomming devices (as soon as I have fully understood what to request, see "I'm still figuring things out" ;) ), but I suspect they are not eager to help with "older" devices like the NS50MU.

BTW did I get the documentation right, that with an UEFI capsule, no plugin like the flashrom plugin is required?

**Reply by superm1** [reply]
That's right. You advertise in the UEFI ESRT what your firmware version is and you publish updates into the ESP and the fwupdx64.efi binary distributed with fwupd calls `UpdateCapsule`.

**Reply by Matombo** [reply]
Another learning question: Can I just pack up arbitrary efi shell code/scripts in an UEFI capsule? aka can I just, in theory, pack up the script from the update zip above myself?

**Reply by Matombo** [reply]
> That's right. You advertise in the UEFI ESRT what your firmware version is and you publish updates into the ESP and the fwupdx64.efi binary distributed with fwupd calls `UpdateCapsule`.

Sorry more noob questions xD: What does ESRT and ESP stand for?

ESP as in the fat32 efi partition?

**Reply by superm1** [reply]
> Another learning question: Can I just pack up arbitrary efi shell code/scripts in an UEFI capsule? aka can I just, in theory, pack up the script from the update zip above myself?

No; it takes an unspecified binary format and uses hooks and code already in the firmware to flash the new one.

> Sorry more noob questions xD: What does ESRT and ESP stand for?

EFI System Resources Table.  Here's a [starter link ](https://learn.microsoft.com/en-us/windows-hardware/drivers/bringup/populating-the-esrt-table)for you on it.

> ESP as in the fat32 efi partition?

Yes

**Reply by Matombo** [reply]
> > Another learning question: Can I just pack up arbitrary efi shell code/scripts in an UEFI capsule? aka can I just, in theory, pack up the script from the update zip above myself?
> 
> No; it takes an unspecified binary format and uses hooks and code already in the firmware to flash the new one.

So nothing one can do without documentation/help from Insyde? ... That's a problem as explained above.

> 
> > Sorry more noob questions xD: What does ESRT and ESP stand for?
> 
> EFI System Resources Table. Here's a [starter link ](https://learn.microsoft.com/en-us/windows-hardware/drivers/bringup/populating-the-esrt-table)for you on it.

Thank you very much

> 
> > ESP as in the fat32 efi partition?
> 
> Yes

**Reply by superm1** [reply]
Yeah you need to talk to Insyde.

**Reply by Matombo** [reply]
Talked to a college who already tried to get Capsule updates for Clevo devices serveral times already: No luck so far and in the forseeable future. So this is hitting a hard block...

**Comment by Matombo** [comment]
Trying out a different device with an AMI BIOS, here i have a Capsule update.

1. get- devices is showing the BIOS as a generic "UEFI ESRT device". Do I have to add the guid of that device first somewhere so that it gets recognized and shown as the BIOS device?
2. The capsule flashes BIOS + EC but not ME, which is just flashed directly, bringing me back to one of my original questions: Can the flashrom plugin do a partial flash (in this case just the ME part)?
3.  Can 2 different plugins flash the same device consecutifly? How encode that in quirks/metainfo.xml? In this case 1. flash ME using flashrom plugin, 2. flash UEFI-capsule

**Reply by hughsie** [reply]
You can add that info to the quirk file, but the best thing to do is to use the metadata -- when fwupd sees metadata for UEFI devices with matching GUIDs it copies the details into the local device.

>  can the flashrom plugin do a partial flash

I guess it can, but we can;t run any nsh script at startup.

> Can 2 different plugins flash the same device consecutifly

I don't think so; although in fwupd I think you've probably got *two* devices under the hood anyway.

**Reply by Matombo** [reply]
> You can add that info to the quirk file, but the best thing to do is to use the metadata -- when fwupd sees metadata for UEFI devices with matching GUIDs it copies the details into the local device.

I'm still a litte bit confused by this:
- `fwupd get-devices` gives me only 2 GUID for the device, one generic "main system firmware" one and one specific to this device, but i don't know how that one is generated and if it is shared on similar devices.
- `fwupdtool hwids` gives me several guid based on the dmi ids of the device, but I seem to only be able to use these in the flashrom.quirks, as when i use them in the metainfo the device is not recignized

How do I devine this device as the BIOS device in the metainfo, and can it be done based in dmi id like with the flashrom plugin?

> 
> > can the flashrom plugin do a partial flash
> 
> I guess it can, but we can;t run any nsh script at startup.
> 
> > Can 2 different plugins flash the same device consecutifly
> 
> I don't think so; although in fwupd I think you've probably got _two_ devices under the hood anyway.

I'm currently talking about another device then in my first post: On this device I have a capsule update, but it leaves out the ME section, that then musst be flashed directly.

Also only parts of the capsule update should be used, which actually again involes an efi binary: In the efi shell you execut an efi binary with one or more flags and the capsule file, then the device reboots and flashes the part of the capsule files specified by the flags. Is this a standardise functionality of capsule files fwupd is capable off? or again something proprietary that would need reverse engeneering).

For reference the hwids and get-devices output of the 3 devices I'm testing with:

The Tuxedo IBS15 Gen6/Clevo NS50MU from the initial post (Insyde BIOS, no capsule available, flash partially direct, partially by ominous efi binary):
[fwupdmgr_get-devices_IBS15Gen6.txt](https://github.com/fwupd/fwupd/files/11422878/fwupdmgr_get-devices_IBS15Gen6.txt)
[fwupdtool_hwids_IBS15Gen6.txt](https://github.com/fwupd/fwupd/files/11422882/fwupdtool_hwids_IBS15Gen6.txt)

Tuxedo Stellaris 15 AMD Gen3/TongFang GMxZGxx (AMI BIOS, capsule available, only parts of capsule should be flashed):
[fwupdmgr_get-devices_Stellaris_15_AMD_Gen3.txt](https://github.com/fwupd/fwupd/files/11422880/fwupdmgr_get-devices_Stellaris_15_AMD_Gen3.txt)
[fwupdtool_hwids_Stellaris_15_AMD_Gen3.txt](https://github.com/fwupd/fwupd/files/11422883/fwupdtool_hwids_Stellaris_15_AMD_Gen3.txt)

Tuxedo Stellaris 15 Intel Gen3/TongFang GMxTGxx (AMI BIOS, capsule available, only parts of capsule should be flashed, ME update as seperate binary that must be flashed directly):
[fwupdmgr_get-devices_Stellaris_15_Intel_Gen3.txt](https://github.com/fwupd/fwupd/files/11422881/fwupdmgr_get-devices_Stellaris_15_Intel_Gen3.txt)
[fwupdtool_hwids_Stellaris_15_Intel_Gen3.txt](https://github.com/fwupd/fwupd/files/11422884/fwupdtool_hwids_Stellaris_15_Intel_Gen3.txt)

**Reply by Matombo** [reply]
The ME for the intel devices is not a seperate device under the hood but lies on the bios chip.

**Reply by Matombo** [reply]
> You can add that info to the quirk file, but the best thing to do is to use the metadata -- when fwupd sees metadata for UEFI devices with matching GUIDs it copies the details into the local device.

Sorry for bumping: How do I do that?

And 2nd question: Can fwupd execute userspace scripts/binaries? Background: For both AMI and Insyde BIOSes official (proprietary) linux flash tools exist. This might be an alternative if I can't get cooperation from the ODM.

**Comment by hughsie** [comment]
> Sorry for bumping: How do I do that?

You just upload the firmware to the LVFS, enable the remote client side, refresh the remote and it magically gets copied to the local device.

> Can fwupd execute userspace scripts/binaries

Unfortunately not; for both technical and legal reasons.

**Reply by Matombo** [reply]
> > Sorry for bumping: How do I do that?
> 
> You just upload the firmware to the LVFS, enable the remote client side, refresh the remote and it magically gets copied to the local device.

Ah sorry: I meant how do I define that in the metainfo.xml?

For the devices with the flashrom plugin, they are automatically labed as the bios device and get uuids assigned based in the dmi strings.

On the devices where the capsule updates should be used, the device has only 2 Guids, one for generic "main system firmware" and one I don't know how it gets generated.

> > Can fwupd execute userspace scripts/binaries
> 
> Unfortunately not; for both technical and legal reasons.

ok

**Reply by hughsie** [reply]
> Ah sorry: I meant how do I define that in the metainfo.xml?

Just like normal; e.g. https://lvfs.readthedocs.io/en/latest/metainfo.html - then things like the LVFS::VersionFormat and name get propagated to the local device.
